core/src/main/kotlin/net/corda/core/internal/TransactionVerifierServiceInternal.kt

```
@@ -29,45 +29,80 @@ fun LedgerTransaction.prepareVerify(extraAttachments: List<Attachment>) = this.i
-        val hashToSignatureConstrainedContracts = verifyConstraintsValidity()
-        verifyConstraints(hashToSignatureConstrainedContracts)
+
+        // 5. Final step is to run the contract code. After the first 4 steps we are now sure that we are running the correct code.
         verifyContracts()
     }
 
-    // TODO: revisit to include contract version information
     /**
-     *  This method may return more than one attachment for a given contract class.
-     *  Specifically, this is the case for transactions combining hash and signature constraints where the hash constrained contract jar
-     *  will be unsigned, and the signature constrained counterpart will be signed.
+     *  This method returns the attachment with the code for each contract.
+     *  It makes sure there is one and only one.
+     *  This is an important piece of the security of transactions.
      */
-    private fun getContractAttachmentsByContract(): Map<ContractClassName, Set<ContractAttachment>> {
+    private fun getUniqueContractAttachmentsByContract(): Map<ContractClassName, ContractAttachment> {
         val contractClasses = allStates.map { it.contract }.toSet()
-        val result = mutableMapOf<ContractClassName, Set<ContractAttachment>>()
 
-        for (attachment in ltx.attachments) {
-            if (attachment !is ContractAttachment) continue
-            for (contract in contractClasses) {
-                if (contract !in attachment.allContracts) continue
-                result[contract] = result.getOrDefault(contract, setOf(attachment)).plus(attachment)
-            }
-        }
+        // For each attachment this finds all the relevant state contracts that it provides.
+        // And then maps them to the attachment.
+        val contractAttachmentsPerContract: List<Pair<ContractClassName, ContractAttachment>> = ltx.attachments
+                .mapNotNull { it as? ContractAttachment }
+                .distinctBy { it.id }
```

Hmm, I had to look up and puzzle for a moment over this `distinctBy` line. It appears to be ignoring the case where the same attachment is attached twice. Should we detect this earlier then throw an exception rather than silently dropping the duplicates here?

---

Duplicates should be harmless as far as I can think of.
But I guess it's better to be cautious about it and throw at some earlier point.